[inductor] remove redundant view/permute earlier

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #171359
* #136702
* __->__ #170649

2 benefits
1. remove these redundancy earlier so that AutoChunker does not need to handle redundant view pairs
2. some attention patterns add redudant views in the pattern on purpose for matching against the compiled graph. By removing these redundancy in the compiled graph earlier, we can simplify those patterns a bit by not injecting redundant views.

test by the existing tests in test_fused_attention.py

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

shunting314 committed:
[inductor] remove redundant view/permute earlier

[ghstack-poisoned]

eellison reviewed (commented):
Should we apply these canonicalization patterns to the patterns generated from register_replacement, as well ?

shunting314 commented:
> Should we apply these canonicalization patterns to the patterns generated from register_replacement, as well ?

I think both should work.  But the change to the attn pattern makes them simpler. So I would still prefer changing those patterns.

On the other hand, we can apply those canonicalization  patterns in the trace function generating patterns as the same time. But I have 2 concerns
1. applying other patterns when generating a more complex pattern seems overall tricky. Maybe it's fine if the applied patterns are simple enough
2. I think it's more helpful if the canonicalization patterns removes redundancies generated due to tracing (rather than due to how the code being traced is written). If we can improve the code being traced to remove those redundancy, I like better to just change the original code.

I'm open to more discussions.

v0i0 commented on torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_21.py:
```diff
@@ -42,47 +42,43 @@
 view_default_1 = CallFunction(aten.view.default, clone_default_1, Ignored(), _users=2)
```
not really feedback, just curious: what are these files doing?

shunting314 replied:
These are like cached compiled patterns. Used to speedup compilation

shunting314 committed:
Update on "[inductor] remove redundant view/permute earlier"


2 benefits
1. remove these redundancy earlier so that AutoChunker does not need to handle redundant view pairs
2. some attention patterns add redudant views in the pattern on purpose for matching against the compiled graph. By removing these redundancy in the compiled graph earlier, we can simplify those patterns a bit by not injecting redundant views.

test by the existing tests in test_fused_attention.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

eellison reviewed (commented):
Since this runs prior to all post grad, and joint graph passes, can we apply this to user patterns as well ? this would just be added as a pass to register_replacement.

shunting314 committed:
Update on "[inductor] remove redundant view/permute earlier"


2 benefits
1. remove these redundancy earlier so that AutoChunker does not need to handle redundant view pairs
2. some attention patterns add redudant views in the pattern on purpose for matching against the compiled graph. By removing these redundancy in the compiled graph earlier, we can simplify those patterns a bit by not injecting redundant views.

test by the existing tests in test_fused_attention.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] remove redundant view/permute earlier"


2 benefits
1. remove these redundancy earlier so that AutoChunker does not need to handle redundant view pairs
2. some attention patterns add redudant views in the pattern on purpose for matching against the compiled graph. By removing these redundancy in the compiled graph earlier, we can simplify those patterns a bit by not injecting redundant views.

test by the existing tests in test_fused_attention.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] remove redundant view/permute earlier"


2 benefits
1. remove these redundancy earlier so that AutoChunker does not need to handle redundant view pairs
2. some attention patterns add redudant views in the pattern on purpose for matching against the compiled graph. By removing these redundancy in the compiled graph earlier, we can simplify those patterns a bit by not injecting redundant views.

test by the existing tests in test_fused_attention.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]
